rep-0149.rst

```diff
@@ -505,10 +491,48 @@ Attributes
   
     "You must give any other recipients of the Work or Derivative Works a copy of this License"
 
+ ``source-files="FILENAME-PATTERN"`` *(optional)*
+
+  A filename pattern using the simplified shell glob syntax specified in Section 6.9 of the `Machine-readable
+  debian/copyright file specification V1.0 <https://www.debian.org/doc/packaging-manuals/copyright-format/1.0/>`_
+  and relative to the ``package.xml`` file.
+
+  The filename pattern specifies the source files this license information refers to. The value
+  ``source-files="*"`` refers to all source files of the package, including source files that are downloaded automatically
+  during the build process - for example in the case of so-called *vendor packages*. If the attribute is not specified,
+  the tag again refers to all source files of the package, including downloaded source files.
+
+  If the filename patterns of multiple license tags match a particular file, the last tag applies to it - following
+  the logic described in Section 6.9 of the `Machine-readable
+  debian/copyright file specification V1.0 <https://www.debian.org/doc/packaging-manuals/copyright-format/1.0/>`_.
+  Consequently, more general tags should be given first.
+
+ ``copyright="COPYRIGHT HOLDER INFORMATION"`` *(optional)*
```

Comment:
I'd like to see an example of the kind of information you'd like to see in here.  In particular, I'm concerned that with files that have multiple copyright holders, this attribute could get extremely long.  Maybe if we introduce the "element" based license as I suggest above, this could be another set of elements.

Reply:
I introduced the `copyright` attribute only for so-called vendor packages, i.e., ROS packages that wrap a non-ROS library and download the source-code of the library during the build from the original location. An example is https://github.com/ros2/tinyxml_vendor/, which downloads the TinyXML library from https://sourceforge.net/projects/tinyxml/. Since the source code of TinyXML is not part of the tinyxml_vendor repository, typical license checker tools will not be able to obtain the copyright information of the TinyXML library. In case of the TinyXML library, the copyright information for the `copyright` attribute would probably be `Original code by Lee Thomason (www.grinninglizard.com)` as stated in the license headers of the source files.
Personally, I have to say that I don't like the current form of vendor packages: I would prefer that the actual source code of the wrapped non-ROS library is directly included in the package. This would make the vendor packages self-contained and thus avoid the need for an internet connection during the build. In this case, the `copyright` attribute would not be required at all since the copyright information could be obtained by standard tools such as licensecheck.

Reply:
>  In case of the TinyXML library, the copyright information for the copyright attribute would probably be Original code by Lee Thomason (www.grinninglizard.com) as stated in the license headers of the source files.

Right, I understand that bit.  The problem is that it is not uncommon to have copyright lines like:

```
Copyright: Contributer A 2010-2012
Copyright: Contributer B 2013-2020
```

Like above, this ends up putting a lot of data into an XML attribute, which just seems weird.

> Personally, I have to say that I don't like the current form of vendor packages: I would prefer that the actual source code of the wrapped non-ROS library is directly included in the package.

We've tried both.  While I also don't like the current form of vendor packages for several reasons, it turns out that they are easier for us to maintain this way.  In particular, we often have to do additional things with CMake to make vendor packages fit into the rest of the system.  With the current vendor packages, we wrap that all into the vendor package itself, basically leaving the original tarball that we download alone.  When we've tried this by forking the upstream repository, we've had to add our own additional patches into the git history.  This works, but updating to a later version becomes tricky with merges.  I guess another way we could try would be to put a *copy* of the sources into our vendor packages (i.e. don't fork the upstream repository), and then periodically update it.  We'll have to think about it.

All of that said, the goal is always to get rid of vendor packages.  With a couple of exceptions, most vendor packages on Ubuntu 22.04 end up being empty.  The vendor packages mostly exist to support Windows, and in the long-term we hope to use another solution there.

Reply:
> I guess another way we could try would be to put a copy of the sources into our vendor packages (i.e. don't fork the upstream repository), and then periodically update it. We'll have to think about it.

I would prefer that way. All vendor packages specify a specific version of the source code to download in the CMakeLists.txt. Thus changing to a newer version of the wrapped library always requires a change in vendor package anyway.

> All of that said, the goal is always to get rid of vendor packages.
In this case, I'll remove my proposal for a copyright attribute from this PR.